### PR TITLE
[FEATURE] Build the manifest-bundle.zip for applications and libraries

### PR DESCRIPTION
**Pull request by petermuessig**
Fixes: https://github.com/SAP/ui5-builder/issues/144

**Thank you for your contribution!** 🙌

To get it merged faster, kindly review the checklist below:

## Pull Request Checklist
- [x] Reviewed the [Contributing Guidelines](https://github.com/SAP/ui5-tooling/blob/master/CONTRIBUTING.md#-contributing-code)
    + Especially the [How to Contribute](https://github.com/SAP/ui5-tooling/blob/master/CONTRIBUTING.md#how-to-contribute) section 
- [x] [No merge commits](https://github.com/SAP/ui5-tooling/blob/master/docs/Guidelines.md#no-merge-commits)
- [x] [Correct commit message style](https://github.com/SAP/ui5-tooling/blob/master/docs/Guidelines.md#commit-message-style)

### CODE REVIEW

**Comment by petermuessig** [comment]
Might be that this is not sufficient when the manifest is created virtually first since the globber only includes sources and not virtual resources...

**Comment by petermuessig** [comment]
The resource globbing should be adopted to also consider virtual resources in the task.

File: lib/types/library/LibraryBuilder.js [hunk 1]
```diff
@@ -67,6 +67,16 @@ class LibraryBuilder extends AbstractBuilder {
 			});
 		});
 
+		this.addTask("generateManifestBundle", () => {
+			const generateManifestBundle = tasks.generateManifestBundle;
+			return generateManifestBundle({
+				workspace: resourceCollections.workspace,
+				options: {
+					namespace: project.metadata.namespace
```
Comment (RandomByte) [inline lib/types/library/LibraryBuilder.js]: Library projects do not have a `metadata.namespace` property. Reason is that they might have multiple namespaces (like sap.ui.core).

I think this still works as the namespace option is only used to:
> `// Root-project only: Remove namespace prefix if given`

See:
https://github.com/SAP/ui5-builder/blob/bb2fcf20526c6072278f9346c97bb184bb23b82b/lib/processors/bundlers/manifestBundler.js#L104-L106

But even if it works, we should make the parameter optional in the JSDoc and do not pass it here as it is always undefined.

Comment (RandomByte) [reply lib/types/library/LibraryBuilder.js]: Actually I think this doesn't work at all. The second parameter of `addBuffer` is the path *inside* the zip file, which must not start with `/`. Have you tested this by any chance?

I'll have a look and try and crate a test for this.

Comment (petermuessig) [reply lib/types/library/LibraryBuilder.js]: Your points have been tackled with the new upload of the commit. I also merged the commits and aligned to the commit message rules.

File: lib/processors/bundlers/manifestBundler.js [hunk 2]
```diff
@@ -107,8 +107,11 @@ module.exports = ({resources, options}) => {
 		});
 		zip.end();
 
+		// TODO: applications are flat / libraries keep their folder structure
+		//       when flattening the libraries this also needs to be adopted or made more flexible
+		const pathPrefix = (options.projectType === "library" ? "/resources/" + options.namespace + "/" : "/");
```
[the remark this answers is not in the document]
Comment (petermuessig) [reply lib/processors/bundlers/manifestBundler.js]: The projectType is more accurate for the moment and the project.type is also used in several other places, so why not here as well?

[the remark this answers is not in the document]
Comment (petermuessig) [reply lib/processors/bundlers/manifestBundler.js]: Ok, I can omit it easily

File: test/lib/builder/builder.js [hunk 3]
```diff
@@ -53,14 +53,16 @@ function cloneProjectTree(tree) {
...
 		const expectedFile = expectedFiles[i];
-		const relativeFile = path.relative(expectedPath, expectedFile);
-		const destFile = path.join(destPath, relativeFile);
-		const currentFileContentPromise = readFile(destFile, "utf8");
-		const expectedFileContentPromise = readFile(expectedFile, "utf8");
-		const assertContents = ([currentContent, expectedContent]) => {
-			assert.equal(currentContent.replace(newLineRegexp, "\n"), expectedContent.replace(newLineRegexp, "\n"));
-		};
-		await Promise.all([currentFileContentPromise, expectedFileContentPromise]).then(assertContents);
+		if (!expectedFile.endsWith(".zip")) {
```
[the remark this answers is not in the document]
Comment (petermuessig) [reply test/lib/builder/builder.js]: There is a separate check for the content of the ZIP file - but we omit to 1..1 compare the ZIP file content in that exec here.

[the remark this answers is not in the document]
Comment (RandomByte) [reply test/lib/builder/builder.js]: This is my point. Since there is a separate check, we should not test the `generateManifestBundle` task here as well. Only testing for the *existence* of  the ZIP files does not sufficiently test the task, so rather not test it at all. Otherwise we cover code without testing it since most of the tasks code is about the *content* of the ZIP file.

File: lib/tasks/bundlers/generateManifestBundle.js [hunk 4]
```diff
@@ -11,25 +12,29 @@ const BUNDLE_NAME = "manifest-bundle.zip";
  * @param {Object} parameters Parameters
  * @param {module:@ui5/fs.DuplexCollection} parameters.workspace DuplexCollection to read and write files
  * @param {Object} parameters.options Options
- * @param {string} parameters.options.namespace Namespace of the application
+ * @param {string} parameters.options.projectName Project name
+ * @param {string} [parameters.options.namespace] Namespace of the application (can be omitted for libraries)
```
[the remark this answers is not in the document]
Comment (petermuessig) [reply lib/tasks/bundlers/generateManifestBundle.js]: Because it is needed and I mixed my brain with the cachebuster info and the manifest bundle creator... :-/ Thanks for your exact review and your patience with me... :-)

**Comment by RandomByte** [comment]
> Is it possible that a library contains multiple manifest.json resources?
> If yes, will they be included into a single bundle or multiple?
> And what is the expected path of the bundle(s) in the dist directory?

To answer my own questions after our brief talk today:
- Yes
- Into a single bundle
- The root directory of the *main* namespace of the library. Similar to where JSDoc output is placed

File: lib/tasks/bundlers/generateManifestBundle.js [hunk 5]
```diff
@@ -11,25 +12,29 @@ const BUNDLE_NAME = "manifest-bundle.zip";
  * @param {Object} parameters Parameters
  * @param {module:@ui5/fs.DuplexCollection} parameters.workspace DuplexCollection to read and write files
  * @param {Object} parameters.options Options
- * @param {string} parameters.options.namespace Namespace of the application
+ * @param {string} parameters.options.projectName Project name
+ * @param {string} [parameters.options.namespace] Namespace of the application/library
```
Comment (RandomByte) [inline lib/tasks/bundlers/generateManifestBundle.js]: Not optional anymore, isnt' it?

File: test/lib/builder/builder.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (RandomByte) [reply test/lib/builder/builder.js]: Still to be discussed

[the remark this answers is not in the document]
Comment (RandomByte) [reply test/lib/builder/builder.js]: Is this change still needed? The builder tests don't produce zip files anymore, now that the task is not executed by default anymore.

Comment (RandomByte) [inline test/lib/builder/builder.js]: Is this change still needed? The builder tests don't produce zip files anymore, now that the task is not executed by default anymore.

File: lib/tasks/bundlers/generateManifestBundle.js (hunk 5 above)
Comment (petermuessig) [reply lib/tasks/bundlers/generateManifestBundle.js]: Done!

File: test/lib/builder/builder.js (hunk 3 above)
Comment (petermuessig) [reply test/lib/builder/builder.js]: Good catch. Removed!
